### DISCUSSION TITLE
Hibernate Reactive: Upserting millions of records

### DISCUSSION THREAD

**Discussion by csotiriou** [quarkusio/quarkus#28586] [Q&A] [answered]
I am using Hibernate Reactive, and I have a package of millions of records spanning 5 tables. The point is that this package needs to be inserted into the database as a single transaction, meaning that if anything goes wrong, we need to roll back. Apart from this procedure, no other writes are being done in the database (it's being updated once per a few months, and all other operations needed are read operations)

For this reason, I figured that putting all of those in a single transaction is going to be terribly inefficient in terms of memory, therefore I was wondering whether I could use the `flush` method for this, meaning that I could open a transaction, have the files inserted as batches of 2000, `flush()` at the end of each batch, and then commit the transaction when everything is done.

I have read the documentation here (https://quarkus.io/guides/hibernate-reactive-panache) and I have the following questions:

- Would using a combination of transactions and` flush()` be the right approach here with Hibernate Reactive?
- if there is an error before the transaction is committed, will the data be rolled back even if there are batches that have been `flush`ed?

Thank you in advance

[a comment by an automated account is not shown]

**Comment by DavideD** [comment, marked as answer]
Yes, flushing the changes to the database won't commit the transaction. So, at the end of all the operations, you can commit or rollback the transaction, and it should work as expected.

One thing to keep in mind is that there is an [open issue ](https://github.com/quarkusio/quarkus/issues/24499) about rollbacking the transaction in case of errors.
The gist of it is that when there is an exception, one must discard the current session and create a new one. Panache doesn't do that at the moment, so you will have to handle that use case. [We discussed some workarounds](https://github.com/quarkusio/quarkus/discussions/23166#discussioncomment-3023637) but it might be easier to handle this use case without Panache.

**Reply by csotiriou** [reply]
Thank you for the reply, and the heads up!

One thing to understand regarding the issue with the rollback, the way I see it from your comment here is that I have two options:
- Drop Panache and use the plain Hibernate Reactive for transactionality
- ... or when the exception is caused, I should call `Arc.container().instance(Mutiny.Session.class).destroy()` in the `onFailure()` handler of Mutiny (in the chain).

Am I getting this correctly?

**Reply by DavideD** [reply]
Yes, that's correct

**Reply by csotiriou** [reply]
Thank you very much for your prompt reply, I appreciate it!
